docs/source/notes/production_ecosystem.rst

```diff
@@ -0,0 +1,131 @@
+taken by individual operators on demand. One can use the same mechanism to do
+always on measurements for any process running PyTorch. It might be useful for
+gathering information about PyTorch workload running in a given process or
+across the entire set of machines.
+
+New callbacks for any operator invocation can be added with
+``torch::autograd::profiler::pushCallback``. Hooks will be called with
+``torch::autograd::profiler::RecordFunction`` struct that describes invocation
+context (e.g. `name`). If enabled, ``RecordFunction::inputs()`` contains arguments
+of the function represented as ``torch::IValue`` variant type. Note, that inputs
+logging is relatively expensive and thus has to be enabled explicitly.
+
+Invoking callbacks adds some overhead, so usually it's useful to just randomly
+sample operator invocations. This can be enabled on per-callback basis with a
+global sampling rate specified by
+`torch::autograd::profiler::setSamplingProbability`.
+
+Potential example might look like:
+
+.. code-block:: cpp
+
+    // Called somewhere in the program beginning
+    void init() {
+        // Sample one in a hundred operator runs randomly
+        torch::autograd::setSamplingProbability(0.01);
+        pushCallback(
+            &onFunctionEnter,
+            &onFunctionExit,
+            /* needs_inputs */ true,
+            /* sampled */ true
+        );
+    }
+
+    void onFunctionEnter(const RecordFunction& fn) {
+        std::cerr << "Before function " << fn.name() 
+                  << " with " << fn.inputs().size() << "inputs" << std::endl;
+    }
+
+    void onFunctionExit(const RecordFunction& fn) {
+        std::cerr << "After function " << fn.name();
+    }
+
+Attaching metadata to saved TorchScript models
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+TorchScript modules can be saved as an archive file that bundles serialized
+parameters and module code as TorchScript (see :meth:`torch.jit.save`). It's
+often convenient to bundle additional information together with the model, for
+example, description of model producer or auxiliary artifacts.
+
+It can be achieved by passing ``_extra_files`` argument to
+:meth:`torch.jit.save` and ``torch::jit::load`` to store and retrieve
+arbitrary binary blobs during saving process. Since TorchScript files are
+regular ZIP archives, extra information gets stored as regular files inside
+archive's ``extra/`` directory.
+
+There's also a global hook allowing to attach extra files to any TorchScript
+archive produced in the current process. It might be useful to tag models with
+producer metadata, akin to JPEG metadata produced by digital cameras. Example
+usage might look like:
+
+.. code-block:: cpp
+
+    SetExportModuleExtraFilesHook([](const script::Module&) {
+        script::ExtraFilesMap files;
+        files["producer_info.json"] = "{\"user\": \"" + getenv("USER") + "\"}";
+        return files;
+    });
+
+
+API usage logging
```

Comment:
move this section above "Attaching metadata to saved TorchScript models" as it flows much better after the operator profiling section